### PR TITLE
Async core APIs in java agent

### PR DESCRIPTION
**Pull request by wu-sheng**
Hi
I am thinking about async core APIs for several months, but no action until now. In the discussion with @BFergerson in #2341 for vert.x scenario, I have some practical ideas. 

cc @adriancole You may be interested about this.

#### Async Span APIs
There is a set of advanced APIs in Span, which work specific for async scenario. When tags, logs, attributes(including end time) of the span
needs to set in another thread, you should use these APIs.

```java
    /**
     * The span finish at current tracing context, but the current span is still alive, until {@link #asyncFinish}
     * called.
     *
     * This method must be called<br/>
     * 1. In original thread(tracing context).
     * 2. Current span is active span.
     *
     * During alive, tags, logs and attributes of the span could be changed, in any thread.
     *
     * The execution times of {@link #prepareForAsync} and {@link #asyncFinish()} must match.
     *
     * @return the current span
     */
    AbstractSpan prepareForAsync();

    /**
     * Notify the span, it could be finished.
     *
     * The execution times of {@link #prepareForAsync} and {@link #asyncFinish()} must match.
     *
     * @return the current span
     */
    AbstractSpan asyncFinish();
```
1. Call `#prepareForAsync` in original context.
1. Propagate the span to any other thread.
1. After all set, call `#asyncFinish` in any thread.
1. Tracing context will be finished and report to backend when all spans's `#prepareForAsync` finished(Judged by count of API execution).

### CODE REVIEW

**Comment by wu-sheng** [comment]
This is my first draft. Please review, try and provide feedback.

**Comment by wu-sheng** [comment]
@liuhaoyang I used to avoid the issue in .net, now, please take a look at this.

**Comment by wu-sheng** [comment]
I think I miss to clear the trace context when stack is empty. Try to fix this.

**Comment by wu-sheng** [comment]
New commit
-----

Finish the context based on `#stopSpan` result, rather than listener. Because in new mode, listener is only being activated/called when across-thread spans finished(ref `span#asyncFinish`).

**Comment by adriancole** [comment]
I think for  people not working in skywalking every day, the change here is because the span apis usually are for one thread right? So, if you want to export some commands where callbacks occur on different threads you need a signal to make that safe. Is this correct?

**Comment by wu-sheng** [comment]
@adriancole 

> I think for people not working in skywalking every day, the change here is because the span apis usually are for one thread right? 

Yes.

> So, if you want to export some commands where callbacks occur on different threads you need a signal to make that safe. Is this correct?

Underlying, Yes. For easy understanding, span(with its context/segment) can't be sent to backend as soon as it finished.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/trace/AbstractTracingSpan.java [hunk 1]
```diff
@@ -322,4 +320,19 @@ public AbstractSpan start(long startTime) {
             refs.add(ref);
         }
     }
+
+    @Override public AbstractSpan prepareForAsync() {
+        context = ContextManager.awaitFinishAsync(this);
```
Comment (BFergerson) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/trace/AbstractTracingSpan.java]: `isInAsyncMode` is never set to true in this class. Did you mean:
```suggestion
        isInAsyncMode = true;
        context = ContextManager.awaitFinishAsync(this);
```

**Comment by wu-sheng** [comment]
@BFergerson Yes. `isInAsyncMode` should be set in that way. Merged your change.

**Pull request merged**